### PR TITLE
translations for assetdefinitions

### PR DESCRIPTION
**Pull request by orthagh**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes

image

### CODE REVIEW

**Comment by orthagh** [comment]
I have reached a point where It's functional.
I gladly take comments on the pr (code orga mainly, but also potential bugs, etc)

**Review by trasher — COMMENTED** [review]
Maybe `getDecodedTranslationsField` and `getPluralFormsForLanguage` should have tests also

**Comment by trasher** [comment]
Adding new definition is broken, I get:
```
  *** MySQL query error:
  SQL: INSERT INTO `glpi_assets_assetdefinitions` (`system_name`, `comment`, `is_active`, `icon`, `capacities`, `profiles`, `translations`, `date_creation`, `date_mod`) VALUES ('TestTranslation', '', '0', '', '[]', '[]', 'Array', '2024-04-05 13:02:51', '2024-04-05 13:02:51')
  Error: CONSTRAINT `glpi_assets_assetdefinitions.translations` failed for `glpi`.`glpi_assets_assetdefinitions`

```

File: src/Asset/AssetDefinition.php [hunk 1]
```diff
@@ -847,6 +942,47 @@ private function validateCapacityArray(mixed $capacities, bool $check_values = t
         return $is_valid;
     }
 
+
+    private function getDecodedTranslationsField(): array
+    {
+        if ($this->fields['translations'] === 'null') {
```
[the remark this answers is not in the document]
Comment (orthagh) [reply src/Asset/AssetDefinition.php:948]: It fixes obviously, but I don't know why it fails for translations, but not for capacities or profiles.

**Comment by orthagh** [comment]
> Adding new definition is broken, I get:

@cedric-anne fixed that (ty)

File: ajax/getPlurals.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ * ---------------------------------------------------------------------
+ */
+
+use Glpi\Asset\AssetDefinition;
+
+include('../inc/includes.php');
+
+header("Content-Type: application/json; charset=UTF-8");
+Html::header_nocache();
+Session::writeClose();
```
Comment (cedric-anne) [inline ajax/getPlurals.php:42]: Are these two instructions mandatory ?

File: templates/pages/admin/assetdefinition/translations.html.twig [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        // empty the old plural inputs
+        $('#translations_panel{{ rand }}').html("");
+
+        // Set the language dropdown value (we redo it for edit case)
+        avoid_language_recursion = true;
+        $('#TranslationModal{{ rand }} select[name="language"]').val(language).trigger('change');
+        avoid_language_recursion = false;
+
+        $.get("{{ path('ajax/getPlurals.php') }}", {
+            id: "{{ item.fields['system_name'] }}",
```
Comment (cedric-anne) [inline templates/pages/admin/assetdefinition/translations.html.twig:163]: This param does not seems to be used.
```suggestion
```

File: ajax/getPlurals.php (hunk 2 above)
Comment (orthagh) [reply ajax/getPlurals.php:42]: probably not, but we don't update anything

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**